File: java-application-maven-archetype/src/main/resources/archetype-resources/src/test/java/MainTest.java
#set( $symbol_pound = '#' )
#set( $symbol_dollar = '$' )
#set( $symbol_escape = '\' )
package ${package};

import static java.nio.charset.StandardCharsets.UTF_8;
import static org.assertj.core.api.Assertions.assertThat;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;
import org.junit.Test;

public class MainTest {

  private static String captureStdout(Runnable runnable) throws Exception {

    PrintStream out = System.out;
    try {
      ByteArrayOutputStream buffer = new ByteArrayOutputStream();
      System.setOut(new PrintStream(buffer, true, "UTF-8"));

      runnable.run();

      return new String(buffer.toByteArray(), UTF_8).trim();
    } finally {
      System.setOut(out);
    }
  }

  @Test
  public void main() throws Exception {
    String actual = captureStdout(() -> Main.main(new String[0]));
    assertThat(actual).endsWith(".Main - Hello, World!");
  }

  @Test
  public void instantiate() {
    new Main();
  }
}

File: kotlin-application-maven-archetype/src/main/resources/archetype-resources/src/test/kotlin/MainTest.kt
#set( $symbol_pound = '#' )
#set( $symbol_dollar = '$' )
#set( $symbol_escape = '\' )
package ${package}

import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import java.io.ByteArrayOutputStream
import java.io.PrintStream
import java.nio.charset.StandardCharsets.UTF_8

private fun captureStdout(callable: () -> Unit): String {
  return System.out.let { out ->
    val buffer = ByteArrayOutputStream()
    try {
      System.setOut(PrintStream(buffer, true, "UTF-8"))

      callable()

      return@let buffer.toByteArray().toString(UTF_8).trim()
    } finally {
      System.setOut(out)
    }
  }
}

class MainTest {

  @Test
  fun testMain() {
    val actual = captureStdout { main(arrayOf()) }
    assertThat(actual).endsWith(".MainKt - Hello, World!")
  }

}
